File: 3678-design-task-manager/design-task-manager.cpp
class TaskManager {
public:
    map<int,set<pair<int,int>>> mp; // priority to set of {userid, taskid}
    map<int,pair<int,int>> t; // task to {priority, userid}
    TaskManager(vector<vector<int>>& tasks) {
        for(auto task:tasks){
            add(task[0],task[1],task[2]);
        }
    }
    
    void add(int userId, int taskId, int priority) {
        mp[priority].insert({taskId, userId});
        t[taskId] = {priority, userId};
    }
    
    void edit(int taskId, int newPriority) {
        auto [oldP, userId] = t[taskId];
        t.erase(taskId);
        mp[oldP].erase({taskId, userId});
        if(mp[oldP].size() == 0)
        mp.erase(oldP);
        add(userId, taskId, newPriority);
    }
    
    void rmv(int taskId) {
        auto [oldP, userId] = t[taskId];
        t.erase(taskId);
        mp[oldP].erase({taskId, userId});
        if(mp[oldP].size() == 0)
        mp.erase(oldP);
    }
    
    int execTop() {
        if(mp.size() == 0)
        return -1;
        auto oldP = (mp.rbegin())->first;
        auto [taskId, userId] = *((mp.rbegin())->second).rbegin();
        t.erase(taskId);
        mp[oldP].erase({taskId, userId});
        if(mp[oldP].size() == 0)
        mp.erase(oldP);
        return userId;
    }
};

/**
 * Your TaskManager object will be instantiated and called as such:
 * TaskManager* obj = new TaskManager(tasks);
 * obj->add(userId,taskId,priority);
 * obj->edit(taskId,newPriority);
 * obj->rmv(taskId);
 * int param_4 = obj->execTop();
 */

File: 3678-design-task-manager/design-task-manager.java
class TaskManager {
    private Map<Integer, Task> taskMap; // taskId -> current Task
    private PriorityQueue<Task> maxHeap; // Max-heap by priority then taskId

    public TaskManager(List<List<Integer>> tasks) {
        taskMap = new HashMap<>();
        maxHeap = new PriorityQueue<>((a, b) -> {
            if (a.priority != b.priority) {
                return Integer.compare(b.priority, a.priority); // Higher priority first
            }
            return Integer.compare(b.taskId, a.taskId); // Larger taskId first
        });
        
        for (List<Integer> task : tasks) {
            addTask(task.get(0), task.get(1), task.get(2));
        }
    }
    
    public void add(int userId, int taskId, int priority) {
        addTask(userId, taskId, priority);
    }
    
    private void addTask(int userId, int taskId, int priority) {
        Task task = new Task(userId, taskId, priority);
        taskMap.put(taskId, task);
        maxHeap.offer(task);
    }
    
    public void edit(int taskId, int newPriority) {
        // Mark old task invalid and create new one
        Task oldTask = taskMap.get(taskId);
        if (oldTask != null) {
            oldTask.invalidate();
        }
        addTask(oldTask.userId, taskId, newPriority);
    }
    
    public void rmv(int taskId) {
        Task task = taskMap.remove(taskId);
        if (task != null) {
            task.invalidate();
        }
    }
    
    public int execTop() {
        while (!maxHeap.isEmpty()) {
            Task task = maxHeap.poll();
            if (task.isValid()) {
                taskMap.remove(task.taskId);
                return task.userId;
            }
        }
        return -1;
    }
}

class Task {
    int userId;
    int taskId;
    int priority;
    boolean valid;
    
    public Task(int userId, int taskId, int priority) {
        this.userId = userId;
        this.taskId = taskId;
        this.priority = priority;
        this.valid = true;
    }
    
    public void invalidate() {
        this.valid = false;
    }
    
    public boolean isValid() {
        return valid;
    }
}